File: ruby/dice_runtime.rb
# -*- mode: ruby; encoding: utf-8; -*-

class DiceNotationRuntime
  attr_reader :is # insn array
  attr_reader :ip # insn pointer, index of is
  attr_reader :ds # data stack
  attr_reader :cs # call stack
  
  def initialize(insn = nil)
    self.reset(insn)
  end

  def reset(insn)
    @is = insn ? insn.dup : nil
    @ds = []
    @cs = []
    @ip = 0
  end

  def run()
    until @is.empty? do
      break unless self.step
    end
    @ds.shift
  end

  def step()
    v = @is[@ip]
    @ip += 1
    p [v, @ds, @cs]
    case
    when v.nil?
      nil
    when v.is_a?(Fixnum)
      @ds << v
    when v.is_a?(Symbol)
      self.send(v)
    else
      raise "Unexpected token: #{v}"
    end
  end

  # ---- helpers

  def with_args(num_of_operands)
    if num_of_operands.respond_to?(:minmax) then
      v = @ds.pop(num_of_operands.max).reverse
      raise "Stack underflow" if v.length < num_of_operands.min
    else
      v = @ds.pop(num_of_operands).reverse
      raise "Stack underflow" if v.length < num_of_operands
    end
    yield(*v)
  end

  # ---- insns

  def lit; @ds << self.with_args(1){|v| v }; end
  def add; @ds << self.with_args((1..2)){|x, y| x.to_i + y.to_i }; end
  def sub; @ds << self.with_args((1..2)){|x, y| x.to_i - y.to_i }; end
  def mul; @ds << self.with_args((1..2)){|x, y| x.to_i * y.to_i }; end
  def div; @ds << self.with_args((1..2)){|x, y| x.to_i / y.to_i }; end
  def max; @ds << self.with_args(2){|x, y| x > y ? x : y}; end
  def min; @ds << self.with_args(2){|x, y| x < y ? x : y}; end
  def roll; @ds << self.with_args(1){|v| Kernel.rand(v) + 1}; end
  def keep_highest
    @ds += self.with_args(2){|keep, num_values|
      values = @ds.pop(num_values)
      values.sort{|x, y| y <=> x }.first(keep)
    }
  end
  alias :push :lit
  def lt; @ds << self.with_args(2){|x, y| (x < y) ? 1 : 0 }; end
  def gt; @ds << self.with_args(2){|x, y| (x > y) ? 1 : 0 }; end
  def eq; @ds << self.with_args(2){|x, y| (x == y) ? 1 : 0 }; end
  def le; @ds << self.with_args(2){|x, y| (x <= y) ? 1 : 0 }; end
  def ge; @ds << self.with_args(2){|x, y| (x >= y) ? 1 : 0 }; end
  def ne; @ds << self.with_args(2){|x, y| (x != y) ? 1 : 0 }; end

  def dup
    @ds << @ds[-1]
  end

  def pop
    @ds.pop
  end

  def goto
    @ip = @ds.pop
  end

  def for
    skip = @ds.pop
    i = @ds.pop
    if i == 0 then
      @ip += skip - 1
    else
      @cs += [@ip, i]
    end
  end

  def next
    i = @cs.pop
    i -= 1
    if i > 0 then
      new_ip = @cs.pop
      @cs += [new_ip, i]
      @ip = new_ip
    else
      @cs.pop # new_ip
    end
  end
end

File: pegjs/dice_runtime.js
module.exports = (function(){

    function DiceRuntime(insn) {
        this.reset(insn);
    }

    DiceRuntime.prototype.reset = function(insn) {
        this.is = insn ? insn : null; // insn array (memory)
        this.ds = [];                 // data stack
        this.cs = [];                 // call stack
        this.ip = 0;                  // insn pointer, index of is
    };

    DiceRuntime.prototype.run = function() {
        while(this.is.length > 0) {
            if(!this.step())
                break;
        }
        return this.ds.shift();
    };

    DiceRuntime.prototype.step = function() {
        var v = this.is[this.ip++];

        if(!v) {
            return false;
        }
        console.log("IP(", this.ip - 1, "): ", v, this.ds, this.cs); // DEBUG

        if(v % 1 === 0) {
            this.ds.push(v);
        } else if(typeof v === "string" || v instanceof String) {
            this.insns[v].call(this, v);
        } else {
            throw "unexpected insn token: " + v;
        }
        return true;
    };

    DiceRuntime.prototype.withArgs = function(arity, callback) {
        var values = [], m, l;
        if(Array.isArray(arity)) {
            m = arity[1]; l = arity[0];
        } else if(arity % 1 === 0) {
            m = l = arity;
        }
        for(var i = 0; i < m && this.ds.length > 0; i++) {
            values.push(this.ds.pop());
        }
        if(values.length < l)
            throw "Stack underflow";
        return callback.apply(this, values);
    };

    DiceRuntime.prototype.push = function(s, v) {
        if(Array.isArray(v)) {
            for(var i = 0, l = v.length; i < l; i++) {
                s.push(v[i]);
            }
        } else {
            s.push(v);
        }
    };

    DiceRuntime.prototype.pop = function(s, n) {
        if(typeof n === "undefined") {
            return s.pop();
        } else {
            var r = [];
            for(var i = 0; i < n; i++) {
                r.unshift(s.pop());
            }
            return r;
        }
    };

    DiceRuntime.prototype.keep = function(callback) {
        return function(keep, num_values) {
            var values = this.pop(this.ds, num_values);
            values.sort(callback);
            return values.slice(0, keep);
        };
    };

    DiceRuntime.prototype.insns = {
        lit: function() { this.ds.push(this.withArgs(1, function(v) { return v; })); },
        add: function() { this.ds.push(this.withArgs([1,2], function(x, y) { return x + y; })); },
        sub: function() { this.ds.push(this.withArgs([1,2], function(x, y) { return x - y; })); },
        mul: function() { this.ds.push(this.withArgs([1,2], function(x, y) { return x * y; })); },
        div: function() { this.ds.push(this.withArgs([1,2], function(x, y) { return x / y; })); },

        max: function() { this.ds.push(this.withArgs(2, function(x, y) { return x > y ? x : y; })); },
        min: function() { this.ds.push(this.withArgs(2, function(x, y) { return x < y ? x : y; })); },

        roll: function() { this.ds.push(this.withArgs(1, function(v) { return Math.floor(Math.random() * v) + 1; })); },
        keep_highest: function() { this.push(this.ds, this.withArgs(2, this.keep(function(x, y) { return y - x; }))); },
        keep_lowest: function() { this.push(this.ds, this.withArgs(2, this.keep(function(x, y) { return x - y; }))); },

        dup: function() { this.ds.push(this.ds.slice(-1)); },
        push: function() { this.insns.lit(); },
        pop: function() { this.ds.pop(); },

        for: function() {
            var skip, i;
            skip = this.ds.pop();
            i = this.ds.pop();
            if(i == 0) {
                this.ip += skip - 1;
            } else {
                this.cs = this.cs.concat(this.ip, i);
            }
        },
        next: function() {
            var i, new_ip;
            i = this.cs.pop();
            if(--i > 0) {
                new_ip = this.cs.pop();
                this.cs = this.cs.concat(new_ip, i);
                this.ip = new_ip;
            } else {
                this.cs.pop();
            }
        }
    };

    var result = {};

    result.DiceRuntime = DiceRuntime;
    
    return result;
})();
